Introduce ResourceSnapshot interface.

jpeach opened the pull request:
Introduce a ResourceSnaphot interface to abstract the type of the snapshot
stored in a SnapshotCache. This allows applications to use SnapshotCache
without also having to use Snapshot. The usefulness of that is that it
is possible to use SnapshotCache with non-Envoy xDS applications.

This fixes #489.

Signed-off-by: James Peach <jpeach@apache.org>

jpeach commented:
/cc @alecholmez @lobkovilya

alecholmez commented on pkg/cache/v3/simple.go:
```diff
@@ -26,6 +26,35 @@ import (
+// can be stored in a SnapshotCache. This enables applications to use the
+// SnapshotCache watch machinery with their own resource types. Most
+// applications will use Snapshot.
+type ResourceSnapshot interface {
```
If this is an interface at this high of a level why not just call it a `Snapshot`? I think that would make more sense and then subsequent structs that implement can name themselves accordingly. I imagine the snapshot itself as the blueprint for a grouping of resources looks like.

Changing this would require we change the original snapshot struct though which I'm not opposed to

jpeach replied:
Yeh I didn't want to rename `Snapshot`, because it's pretty disruptive and I also didn't have a good suggestion for a new name :)

alecholmez replied:
Hm, what are your thoughts on if we were to deal with that upfront cost? You think the API would be easier to use/consume?

jpeach replied:
I'm not sure that there are big wins to the API from renaming. It's probably a bit nicer, but not sure it's really a big deal. If we have a a larger goal to separate out the envoy protobuf dependencies, maybe we would think of it as part of that.

jpeach replied:
@alecholmez I'm  still inclined to keep the current `Snapshot` name as is for now.

akshaysngupta commented:
Any update on this ?

jpeach commented:
> Any update on this ?

I plan to rebase and update this in the next few days.

jpeach commented:
@alecholmez I rebased and added a basic test, but otherwise decided not to change anything. I'm thinking that we may want to remove `(ResourceSnapshot)ConstructGetVersionMap()` and just build it on demand, but it's probably better to do that in a separate PR.

alecholmez reviewed (approved):
I think this is good to go in
